### DISCUSSION TITLE
Generic `Deconstruct` method

### DISCUSSION THREAD

**Discussion by FaustVX** [dotnet/csharplang#7095] [Language Ideas] [answered]
In this test cases 
```C#
ReadOnlySpan<char> span = "Hello";
_ = span is (int i);
_ = span is (float i0);
_ = span is (var i1); // i1 is ROSpan<char>
_ = span is (value: int i2);
_ = span is (value: var i3);
_ = span is (f: float i4); // i4 is float
_ = span is (f: var i5); // i5 is float
_ = span is (int i6, _);
_ = span is (var i7, _);
_ = span is (i: int i8, _);
_ = span is (i: var i9, _);

static class Ext
{
    public static void Deconstruct<T>(this ReadOnlySpan<char> span, [MaybeNull] out T value)
    where T : ISpanParsable<T>
    {
        _ = T.TryParse(span, default, out value);
    }
    public static void Deconstruct(this ReadOnlySpan<char> span, [MaybeNull] out float f)
    {
        _ = float.TryParse(span, default, out f);
    }
    public static void Deconstruct<T>(this ReadOnlySpan<char> span, [MaybeNull] out T i, out bool _)
    where T : ISpanParsable<T>
    {
        _ = T.TryParse(span, default, out i);
    }
}
```
[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCecAmIA1AD4ACATAIwCwAUGQAwAEZVAdACJQQDmMAewDOCKAGMh7AMID8qAIIwIAGxxCoQgNz16AJVQR8AeRiqAygAcIMADxiAFhHQA+ZkKsxmAXmYAiABKoysoCvtp0APrebh7MGswAFLAIcQCU4VE+7tZxQolYIRApUIzp9JkxOfEJyE5xVOnMAPRN9bnMukaW1naOLuXR2Z7VtcoArqggcTDFFGWRg7EjKhNTtehxAMzzFUPtCVhTBQJFcQAsjS3n7cdFA1lLeQdrdVAArJet7zeFCPeVwyeyTiADY4MwIjtFlUnus4gB2cGQjLQwGJKBTYFQAAcSKhDxh6JeGygAE48eF6GwQawKMwAKIADz+dAA3vRmJzWJtWFQaWQzsxOKgxAIYCJ0GMxAgbAAVZwJBD2eL6QwmcweXpOVxDcEAbQAshAcMBUAA5MbBAC6zAEYxSsuYowmqQ5XIA7vYMKhmI6pgBJbowAAKTiEEGAylQcucbs57LoXKTEOisvYsvQOFD6CEqASuuYciwEEtCHBdpSztQ8yTAF849zefzBcLReKEJLpYrlXlVcZTDgg1qXAD9UaTebLcobRXmLcUlhXYmuQnk1yKvP05ns7n8x5wUWS8oy7b7XOa1z68vOWQedTWC2RWKJVKZfLuyqDP2NT0HNrR8whrGqaFrWqeDpxOWZ7AAIAjKBCS5Jp63q+swAZBjuEZRjGDarmuKY+GmGZZmGeYFoepZQcUF6cletZAA)

Only 3 cases compile. I even tried with 2 `out` parameters (just in case), but that doesn't change.
`i1` works (of course).
And only `i4` and `i5` works (only if I type the parameter name before `f:`).

So I'm wondering why generic `Deconstruct` doesn't work, and how to make it work (if possible, without to write a method for every type I know)

And as a side question, why the compiler can't use the correct `Deconstruct` method when the type is provided (for `i0`)

**Comment by HaloFour** [comment, marked as answer]
Deconstruction overloading is based only on arity.

**Reply by FaustVX** [reply]
It's not possible to "bump" the compiler to support that for a future version ?

**Reply by HaloFour** [reply]
If the team decides to revisit deconstruction overloading, sure.

**Reply by FaustVX** [reply]
I really hope they will

**Reply by FaustVX** [reply]
> And as a side question, why the compiler can't use the correct `Deconstruct` method when the type is provided (for `i0`) ?

So, it's the same answer ?
